lib/net/ldap/connection.rb

```diff
@@ -28,6 +30,54 @@ def initialize(server)
     yield self if block_given?
   end
 
+  # Internal: Connect to the host and port.
+  #
+  # Accepted options:
+  # - host: the hostname or IP address String of the server to connect to.
+  # - port: the port Integer to connect to.
+  # - timeout: the Integer seconds to wait before timing out, or `nil` for no
+  #            timeout (the default).
+  #
+  # This connects to the specified server using non-blocking socket methods.
+  # This allows us to specify our own timeout to prevent blocking.
+  #
+  # Returns a connected Socket object.
+  def connect(options)
+    host, port = options[:host], options[:port]
+    timeout    = options[:timeout]
+
+    addr = Socket.getaddrinfo(host, nil)
+    ip   = addr[0][3]
+    sockaddr = Socket.sockaddr_in(port, ip)
+
+    socket = Socket.new(Socket.const_get(addr[0][0]), Socket::SOCK_STREAM, 0)
+    socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
+
+    begin
+      # start connection attempt
+      socket.connect_nonblock(sockaddr)
+    rescue IO::WaitWritable
+      if IO.select(nil, [socket], nil, timeout)
+        # validate connection succeeded
+        begin
+          socket.connect_nonblock(sockaddr)
+        rescue Errno::EISCONN
+          # already connected, no action necessary
+        rescue Exception => e
+          # unexpected error
+          socket.close
+          raise Net::LDAP::LdapError, "Connection to #{host}:#{port} (#{ip}) failed: (#{e.class}) #{e.message}"
```

Comment:
I think new exception class inheriting `Net::LDAP::Error` should be defined for this error.